modules/http-api/src/main/java/org/mule/runtime/http/api/client/ws/WebSocketCallback.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.http.api.client.ws;
+
+import org.mule.api.annotation.Experimental;
+import org.mule.runtime.api.metadata.TypedValue;
+import org.mule.runtime.http.api.ws.WebSocket;
+import org.mule.runtime.http.api.ws.WebSocketCloseCode;
+
+import java.io.InputStream;
+
+/**
+ * A Callback to receive outbound {@link WebSocket} events.
+ * <p>
+ * Implementations <b>MUST</b> be thread safe
+ *
+ * @since 4.1.5 as experimental
+ */
+@Experimental
+public interface WebSocketCallback {
```

Comment:
How would the callback relate to the future? Particularly in regards to `onConnect`.

Reply:
it doesn't. Grizzly completes the future after these methods are executed. It's a good thing that these callbacks are decoupled from the threading model